app/domain/authentication/authn_k8s/execute_command_in_container.rb

```diff
@@ -0,0 +1,158 @@
+require 'timeout'
+
+module Authentication
+  module AuthnK8s
+
+    ExecuteCommandInContainer ||= CommandClass.new(
+      dependencies: {
+        timeout:                       ENV['KUBE_EXEC_COMMAND_TIMEOUT'],
+        websocket_client:              WebSocket::Client::Simple,
+        ws_client_event_handler_class: WebSocketClientEventHandler,
+        message_log_class:             MessageLog,
+        validate_message:              MessageLog::ValidateMessage.new,
+        logger:                        Rails.logger
+      },
+      inputs:       %i(k8s_object_lookup pod_namespace pod_name container cmds body stdin)
+    ) do
+
+      DEFAULT_TIMEOUT_SEC = 5
+
+      extend Forwardable
+      def_delegators :@k8s_object_lookup, :kube_client
+
+      def call
+        init_ws_client
+        add_websocket_event_handlers
+        wait_for_close_message
+        verify_error_stream_is_empty
+        websocket_messages
+      end
+
+      private
+
+      def init_ws_client
+        ws_client
+        ws_client_event_handler
+      end
+
+      def ws_client
+        @ws_client ||= @websocket_client.connect(server_url, headers: headers)
+      end
+
+      def ws_client_event_handler
+        @close_event_queue       = Queue.new
+        @ws_client_event_handler ||= @ws_client_event_handler_class.new(
+          close_event_queue: @close_event_queue,
+          ws_client:         ws_client,
+          stdin:             @stdin,
+          body:              @body,
+          pod_name:          @pod_name,
+          validate_message:  @validate_message,
+          message_log:       @message_log_class.new,
+          logger:            @logger
+        )
+      end
+
+      def add_websocket_event_handlers
+        # We need to set this as if we'll use @ws_client_event_handler inside
+        # the curly brackets it will look for such a member in the websocket_client
+        # class
+        ws_client_event_handler = @ws_client_event_handler
+        main_thread_tags        = @logger.formatter.current_tags
+        logger                  = @logger
+
+        ws_client.on(:open) do
+          # Add log tags (origin, thread id, etc.) to sub-thread as they are not
+          # passed automatically. We append the sub-thread id to the main one so
+          # we can easily know the flow from the logs and connect between the threads
+          tid             = syscall(186)
+          sub_thread_tags = main_thread_tags.map do |x|
+            x.start_with?("tid=") ? "#{x}=>#{tid}" : x
+          end
+          logger.formatter.current_tags.replace sub_thread_tags
+          ws_client_event_handler.on_open
+        end
+        ws_client.on(:message) { |msg| ws_client_event_handler.on_message(msg) }
+        ws_client.on(:close) { ws_client_event_handler.on_close }
+        ws_client.on(:error) { |err| ws_client_event_handler.on_error(err) }
+      end
+
+      def wait_for_close_message
+        Timeout.timeout(timeout) { @close_event_queue.pop }
```

Comment:
Something I didn't have time to do over the weekend but is definitely worth doing: Inject `Timeout` as a dependency.  Among other things, this allows us to play God with time in our unit tests, so that the UTs themselves can still execute instantly.  That is, in our test double, we can happily interpret "5 seconds" as "5 milliseconds".